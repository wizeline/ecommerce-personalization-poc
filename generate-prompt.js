import { PROMPT_TEMPLATE } from "./constants"

const stringify = (text) => JSON.stringify(text, null, 2)

export const generatePrompt = ({ productsInfo, usersHistoryPurchaseInfo, userInfo }) => {

    return PROMPT_TEMPLATE
        .replace("JSON_DE_PRODUCTOS", stringify(productsInfo))
        .replace("JSON_DE_HISTORIAL_DE_COMPRAS", stringify(usersHistoryPurchaseInfo))
        .replace("IDS_DE_USUARIOS", userInfo.map(({ id }) => id).join(","));
}