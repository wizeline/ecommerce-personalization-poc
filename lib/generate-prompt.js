import fs from "fs";

export const generatePrompt = ({
  usersHistoryPurchaseInfo,
}) => {
  const profiles = fs.readFileSync("types_of_users.txt", 'utf8');
  const purchaseHistory = usersHistoryPurchaseInfo.map(({ description }) => `* ${description}\n\n`).join("");
  const template = `The following list is a purchase history that includes the descriptions of all the products a user has bought:

  PRODUCTS:
  ${purchaseHistory}
  
  Also, we have a list of customer profiles:

  CUSTOMER PROFILES:
  ${profiles}

  Please, give me the user profile number that best match the user purchase history:
  `;
  
  return template;
};