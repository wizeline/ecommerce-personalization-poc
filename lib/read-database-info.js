import sqlite3 from "sqlite3";
import { Op } from 'sequelize';
import { DB_PATH } from "./constants.js";

import {
  Category,
  InfoCategory,
  Info,
  User,
  UserCategory,
} from "./sequalize-database-models.js"

function readSQLiteData(dbPath, query, callback) {
  const db = new sqlite3.Database(dbPath, sqlite3.OPEN_READONLY, (err) => {
    if (err) {
      console.error(err.message);
      return;
    }
  });

  db.all(query, [], (err, rows) => {
    if (err) {
      throw err;
    }
    db.close();
    callback(rows);
  });
}

export const readData = (query) =>
  new Promise((resolve, reject) => {
    readSQLiteData(DB_PATH, query, (datos) => {
      try {
        resolve(datos);
        // resolve(JSON.stringify(datos));
      } catch (err) {
        reject(err);
      }
    });
  });

export const readProductInfo = () => Info.findAll({ order: ['product_id'], limit: 150, raw: true });
export const readUserHistoryPurchaseInfo = (userId) => User.findOne({ where: { user_id: userId } });
export const readUserInfo = () => User.findAll({ limit: 10 });
export const readUserCategory = (userId) => User.findOne({ where: { user_id: userId }, include: Category });
export const readCategories = () => Category.findAll({ raw: true });
export const setProductCategories = async ({ product_id }, category_ids, why) => {
  await InfoCategory.destroy({ where: { product_id } });
  await InfoCategory.bulkCreate(category_ids.map((category_id) => ({
    product_id,
    category_id,
    why,
  })));
}
export const readProductsWithoutCategory = async () => readProductInfo()
  .then((infos) => {
    const product_ids_to_consider = infos.map(({ product_id }) => product_id);

    return InfoCategory.findAll({ where: { product_id: { [Op.in]: product_ids_to_consider } }, attributes: ['product_id'] })
      .then((infoCategories) => infoCategories.map(({ product_id }) => product_id))
      .then((existingProductIdWithCategory) => Info.findAll({
        where: {
          product_id: {
            [Op.in]: product_ids_to_consider,
            [Op.notIn]: existingProductIdWithCategory
          }
        },
        raw: true
      }));
  });

/**
 * Gets all the users that don't have a category associated
 * and have a purchase history of at least one product
 * @returns 
 */
export const readCandidateUsers = async () => {
  const users = await User.findAll({
    limit: 10,
    include: [Info, Category]
  })
  return users.filter((user) => user.infos.length > 0 && user.categories.length === 0);
};

export const setUserCategory = async (user_id, category_id, why) => {
  console.log(user_id, category_id, why);
  // await UserCategory.destroy({ where: { user_id } });
  await UserCategory.create({
    user_id,
    category_id,
    why,
  })
}