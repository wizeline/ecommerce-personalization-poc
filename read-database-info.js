import sqlite3 from "sqlite3";
sqlite3.verbose();

import {
  DB_PATH,
  PRODUCT_QUERY,
  USER_HISTORY_PURCHASE_QUERY,
  USER_QUERY,
} from "./constants.js";

function readSQLiteData(dbPath, query, callback) {
  const db = new sqlite3.Database(dbPath, sqlite3.OPEN_READONLY, (err) => {
    if (err) {
      console.error(err.message);
      return;
    }
  });

  db.all(query, [], (err, rows) => {
    if (err) {
      throw err;
    }
    db.close();
    callback(rows);
  });
}

export const readData = (query) =>
  new Promise((resolve, reject) => {
    readSQLiteData(DB_PATH, query, (datos) => {
      try {
        resolve(datos);
        // resolve(JSON.stringify(datos));
      } catch (err) {
        reject(err);
      }
    });
  });

export const readProductInfo = () => readData(PRODUCT_QUERY);
export const readUserHistoryPurchaseInfo = () =>
  readData(USER_HISTORY_PURCHASE_QUERY);
export const readUserInfo = () => readData(USER_QUERY);
