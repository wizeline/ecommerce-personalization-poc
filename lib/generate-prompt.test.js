import { generatePrompt } from "./generate-prompt"
import { readProductInfo, readUserHistoryPurchaseInfo, readUserInfo } from "./read-database-info.js";

import fs from "fs";



describe("generate prompt correctly", () => {

    it("create a prompt including product info, history of purchase and id of users", () => {
        expect(generatePrompt({
            productsInfo: [{
                name: "a pair of shoes",
                description: "this is a glancy pair of shoes",
                id: "P0023"
            }], usersHistoryPurchaseInfo: [
                {
                    "UX0034": ["P0023", "P0024", "P0024"]
                }
            ], userInfo: [{ id: "UX0034" }, { id: "UX0035" }, { id: "UX0036" }]
        })).toMatchSnapshot();
    });

    it.skip("generates proper output for a real example", async () => {
        const info = {
            productsInfo: await readProductInfo(),
            usersHistoryPurchaseInfo: await readUserHistoryPurchaseInfo(),
            userInfo: await readUserInfo(),
        }

        const prompt = generatePrompt(info);
        fs.writeFileSync("prompt.txt", prompt)
        expect(prompt).toMatchSnapshot();
    })
})