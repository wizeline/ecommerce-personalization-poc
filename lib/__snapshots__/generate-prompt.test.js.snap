// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate prompt correctly create a prompt including product info, history of purchase and id of users 1`] = `
"este es el JSON de un listado de productos de una tienda

[
  {
    "name": "a pair of shoes",
    "description": "this is a glancy pair of shoes",
    "id": "P0023"
  }
]

Este JSON que aparece a continuación es un objeto donde cada key representa a un usuario de la tienda y como value, tiene un array con cada identificador de la lista de productos anterior que ha comprado

[
  {
    "UX0034": [
      "P0023",
      "P0024",
      "P0024"
    ]
  }
]

Para los usuarios con ID UX0034,UX0035,UX0036 generame un JSON con un objeto en la respuesta, donde la key sea el id de usuario y el value un array con tres identificadores de producto que le podríamos recomendar en base a su historial de compras.

Respóndeme exclusivamente el JSON que te pido."
`;
